awscli/botocore/utils.py

```diff
@@ -3305,6 +3339,231 @@ def fetch_token(
+            f'{urllib.parse.urlencode(query_params)}'
+            f'&code_challenge={self.code_challenge[:-1]}'  # trim final '='
+        )
+
+    def _get_new_token(self, start_url, session_name, registration_scopes):
+        registration = self._registration(
+            start_url,
+            session_name,
+            registration_scopes,
+        )
+
+        expected_state = uuid.uuid4()
+
+        authorization_uri = self._get_authorization_uri(
+            registration['clientId'],
+            registration_scopes,
+            expected_state)
+
+        # Even though there's just one uri, this matches the inputs
+        # for the device code flow so that we can reuse the browser handlers
+        authorization_args = {
+            'verificationUri': authorization_uri,
+            'verificationUriComplete': authorization_uri,
+            'userCode': None
+        }
+
+        # Open/display the link, then block until the redirect uri is hit and
+        # the auth code is retrieved
+        self._on_pending_authorization(**authorization_args)
+        auth_code, state = self._auth_code_fetcher.get_auth_code_and_state()
+
+        if auth_code is None:
+            raise AuthorizationCodeLoadError(
+                error_msg='Failed to retrieve an authorization code.'
+            )
+
+        if state != expected_state:
+            raise AuthorizationCodeLoadError(
+                error_msg='State parameter does not match expected value.'
+            )
+
+        return self._create_token_(start_url, registration, auth_code)
+
+    def _create_token_(self, start_url, registration, auth_code):
+        try:
+            response = self._client.create_token(
+                grantType='authorization_code',
+                clientId=registration['clientId'],
+                clientSecret=registration['clientSecret'],
+                redirectUri=self._auth_code_fetcher.redirect_uri_with_port(),
+                codeVerifier=self.code_verifier,
+                code=auth_code
+            )
+            expires_in = datetime.timedelta(seconds=response['expiresIn'])
+            token = {
+                'startUrl': start_url,
+                'region': self._sso_region,
+                'accessToken': response['accessToken'],
+                'expiresAt': self._time_fetcher() + expires_in,
+                # Cache the registration alongside the token
+                'clientId': registration['clientId'],
+                'clientSecret': registration['clientSecret'],
+                'registrationExpiresAt': registration['expiresAt'],
+            }
+            if 'refreshToken' in response:
+                token['refreshToken'] = response['refreshToken']
+            return token
+        except self._client.exceptions.InvalidGrantException as error:
+            print(error)
+        except self._client.exceptions.ExpiredTokenException:
+            raise PendingAuthorizationExpiredError()
```

Comment:
We're doing this today in the device grant handler:
https://github.com/aws/aws-cli/blob/c8e3002ced34a77f52cb31df9298b0129cd95ebd/awscli/botocore/utils.py#L3260-L3261

I imagine it was so we can provide the additional instructions to try again, since this error is more timing based than a result of invalid inputs. I copied forward that pattern here.